vmspawn: add features required by mkosi

sam-leonard-ct opened the pull request:
In order to add a vmspawn verb to mkosi sd-vmspawn needs to support some more features natively.

Mkosi draft PR: https://github.com/systemd/mkosi/pull/2037

Features required:
- [x] Software Trusted Platform Module (swtpm)
- [x] Kernel configuration
- [x] Synthesise a root= kernel argument when direct kernel booting
- [x] Initrd configuration
- [x] Network device management (-nic argument for qemu)
- [X] Allow passing in fds for /dev/kvm and /dev/vhost-vsock

bluca commented:
This should use a transient unit for swtpm, and tie its lifetime to the vm being spawned

bluca commented:
Have a look at src/run/ which implements systemd-run

sam-leonard-ct commented on src/vmspawn/vmspawn.c:
```diff
@@ -438,6 +461,94 @@
+                        NULL,
+                        &sock_fd, 1, /* pass the socket fd to swtpm */
+                        FORK_CLOEXEC_OFF,
+                        &child_pid);
```
Oop, will default the pid to zero.

sam-leonard-ct commented:
I'm going to finish of the kernel command line task then change the swtpm implementation to one that uses transient units.

sam-leonard-ct commented:
Added --qemu-net=user|tap|none for configuring the qemu nic argument.

sam-leonard-ct commented:
Okay I've managed to sort out the transient unit things.
swtpm is now started as a socket/service pair using sd-bus messages.
A transient scope unit tracks the lifetime of the vmspawn process, all started services are bound to this scope.

bluca commented on src/vmspawn/vmspawn-scope.c:
```diff
@@ -0,0 +1,359 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "vmspawn-scope.h"
```
sort alphabetically

sam-leonard-ct replied:
Fixed in ffc1867558.

bluca commented on src/vmspawn/vmspawn-scope.c:
```diff
@@ -0,0 +1,359 @@
+#include "macro.h"
+#include "process-util.h"
+#include "random-util.h"
+#include "sd-bus.h"
```
sd-bus is a public header, so sort in a separate block before internal headers

sam-leonard-ct replied:
Fixed in ffc1867558.

bluca commented on src/vmspawn/vmspawn-scope.c:
```diff
@@ -0,0 +1,359 @@
+#include "unit-def.h"
+#include "unit-name.h"
+#include "strv.h"
+#include <stdio.h>
```
external headers in a separate block at the very top

sam-leonard-ct replied:
Fixed in ffc1867558.

bluca commented on src/vmspawn/vmspawn-scope.c:
```diff
@@ -0,0 +1,359 @@
+                return log_error_errno(r, "Could not watch job: %m");
+
+        if (asprintf(&unmangled, "sd-vmspawn-%"PRIu32"-%s", random_u32(), machine_name) < 0)
+                return -ENOMEM;
```
log_oom()

sam-leonard-ct replied:
Fixed in ffc1867558.

bluca commented on src/vmspawn/vmspawn-scope.c:
```diff
@@ -0,0 +1,359 @@
+
+static int message_add_commands(sd_bus_message *m, const char *exec_type, char ***commands, size_t n_commands) {
+        int r;
+
```
assert(m)

sam-leonard-ct replied:
Fixed in ffc1867558.

bluca commented on src/vmspawn/vmspawn-util.c:
```diff
@@ -33,6 +33,19 @@ OvmfConfig* ovmf_config_free(OvmfConfig *config) {
 }
 
+int parse_qemu_network_stack(const char *s, QemuNetworkStack *net_stack) {
+        if (streq(optarg, "user"))
```
assert(s)
assert(net_stack)

bluca replied:
also did you mean s here instead of optarg?

sam-leonard-ct replied:
Fixed in ffc1867558.

YHNdnzj replied:
Still not fixed - please use `s` instead of `optarg`.

bluca commented on src/vmspawn/vmspawn.c:
```diff
@@ -27,13 +26,15 @@
 #include "tmpfile-util.h"
 #include "vmspawn-settings.h"
 #include "vmspawn-util.h"
+#include "vmspawn-scope.h"
```
sort alphabetically

sam-leonard-ct replied:
Fixed in ffc1867558.

bluca commented on src/vmspawn/vmspawn.c:
```diff
@@ -621,21 +744,76 @@ static int run_virtual_machine(void) {
+                        /* only bail if the user asked for a swtpm */
+                        if (arg_qemu_swtpm > 0)
+                                return log_error_errno(r, "Failed to start swtpm: %m");
+                        log_debug_errno(r, "Failed to start swtpm: %m");
```
"Failed to start swtpm, ignoring:"

sam-leonard-ct replied:
Fixed in ffc1867558.

bluca commented on src/vmspawn/vmspawn.c:
```diff
@@ -621,21 +744,76 @@ static int run_virtual_machine(void) {
+                                if (r < 0)
+                                        return log_oom();
+                        } else
+                                log_warning("Cannot append extra args to kernel cmdline, native architecture doesn't support SMBIOS");
```
"SMBIOS, ignoring"

sam-leonard-ct replied:
Fixed in ffc1867558.

sam-leonard-ct commented on src/vmspawn/vmspawn.c:
```diff
@@ -621,21 +744,76 @@ static int run_virtual_machine(void) {
+                if (arg_qemu_kernel) {
+                        r = strv_extend_strv(&cmdline, STRV_MAKE("-append", kcl), /* filter_duplicates= */ false);
+                        if (r < 0)
+                                log_oom();
```
Fixed in ffc1867558.

sam-leonard-ct commented:
This PR now includes the changes from https://github.com/systemd/systemd/pull/30317 as well.

poettering commented on man/systemd-vmspawn.xml:
```diff
@@ -117,6 +117,21 @@
       </varlistentry>
 
+      <varlistentry>
+        <term><option>--qemu-swtpm=</option><replaceable>BOOL</replaceable></term>
```
why prefix this switch name with "qemu" btw?

I'd like to abstract the user virtualizer a bit, hence where things aren't inherently qemu specific I'd give this a generic name.

moreover, there are multiple vtpm implementations areound, swtpm is just one, I'd also abstract this in the name a bit, and thus name this switch with a generic "vtpm", which gives us some liberty to swap out what is backed by.

sam-leonard-ct replied:
As with many of these I chose the name because that is what mkosi calls the argument.
I'll aim to make more generic names in future.
Fixed in 37e75950ff.

poettering commented on src/vmspawn/vmspawn-scope.c:
```diff
@@ -0,0 +1,383 @@
+                            3,
+                            "Description", "s",  description,
+                            "AddRef",      "b",  1,
+                            "PIDs",        "au", 1, getpid_cached(),
```
please use pidfds for this now. see how newer nspawn does this.

sam-leonard-ct replied:
I tried this originally but my systemd version didn't support it so I assumed I had it wrong. I've now added a similar try with pidfd then without on failure that nspawn does.
Fixed in 37e75950ff.

poettering commented on src/vmspawn/vmspawn-scope.c:
```diff
@@ -0,0 +1,383 @@
+        if (!description)
+                return log_oom();
+
+        r = bus_call_method(bus, bus_systemd_mgr, "StartTransientUnit", &error, &reply,
```
we usually put the first line break right after the `(` of a function call

sam-leonard-ct replied:
Fixed in 37e75950ff.

poettering commented on src/vmspawn/vmspawn-scope.c:
```diff
@@ -0,0 +1,383 @@
+
+        r = unit_name_mangle_with_suffix(unmangled, "as scope name", 0, ".scope", &scope);
+        if (r < 0)
+                return r;
```
missing logging

sam-leonard-ct replied:
Fixed in 37e75950ff.

poettering commented on src/vmspawn/vmspawn-scope.c:
```diff
@@ -0,0 +1,383 @@
+
+        r = unit_name_mangle_with_suffix(service_name, "as service", 0, ".service", &unit_name);
+        if (r < 0)
+                return r;
```
logging

sam-leonard-ct replied:
Fixed in 37e75950ff.

poettering commented on src/vmspawn/vmspawn-scope.c:
```diff
@@ -0,0 +1,383 @@
+        if (r < 0)
+                return bus_log_create_error(r);
+
+        r = sd_bus_message_append(m, "ssa(sv)",
```
as above

sam-leonard-ct replied:
Fixed in 37e75950ff.

poettering commented on src/vmspawn/vmspawn.c:
```diff
@@ -426,20 +437,83 @@ static int cmdline_add_vsock(char ***cmdline, int vsock_fd) {
+
+        *ret_state_tempdir = TAKE_PTR(state);
+
+        return 0;
```
hmm so if this only has a single parameter we could also make this a static template service, no? no need for the transient stuff.

i.e. a pair of regular template units vtpm@.socket + vptm@.service we'd just write to /usr/lib/systemd/system/ like any other unit.

The instance field would then be set to the full path of the state, using unit_name_path_escape() so that it can be referenced via %f in the template.

sam-leonard-ct replied:
How would this template the scope? From what I understand we need to use BindsTo in the unit otherwise they won't be stopped when vmspawn stops.

poettering replied:
They can to BindsTo=%i.scope?

sam-leonard-ct replied:
How would you tell virtiofsd what directory to host?

poettering commented on man/systemd-vmspawn.xml:
```diff
@@ -132,6 +132,18 @@
       </varlistentry>
 
+      <varlistentry>
+        <term><option>--qemu-kernel=</option><replaceable>PATH</replaceable></term>
```
same here, i#d drop the "qemu" prefix. 

in fact we should do that for most other options, unless they directly expose a qemu-specific concept that makes no sense if we ever supported some other backend one day. (not saying that we ever actually want to support any other backend, but i think we should keep the options open at least in the interface)

sam-leonard-ct replied:
Fixed in 37e75950ff.

poettering commented on man/systemd-vmspawn.xml:
```diff
@@ -137,13 +137,25 @@
       </varlistentry>
 
+      <varlistentry>
+        <term><option>--qemu-initrd=</option><replaceable>PATH</replaceable></term>
```
ditto

sam-leonard-ct replied:
Fixed in 37e75950ff.

poettering commented on src/vmspawn/vmspawn.c:
```diff
@@ -57,6 +61,7 @@ STATIC_DESTRUCTOR_REGISTER(arg_image, freep);
 STATIC_DESTRUCTOR_REGISTER(arg_machine, freep);
 STATIC_DESTRUCTOR_REGISTER(arg_qemu_smp, freep);
 STATIC_DESTRUCTOR_REGISTER(arg_qemu_kernel, freep);
+STATIC_DESTRUCTOR_REGISTER(arg_qemu_initrd, freep);
```
shouldn't this be an strv? i.e. kernels are fine with multiple initrds. does qemu support that? if so, then we should do that too. After all we already use multiple initrds for passing through creds, and sysexts and all that stuff. seems odd to not support this here if we make use of it already and qemu supports it too

sam-leonard-ct replied:
I've implemented this now, I'm not 100% sure the arguments to qemu are correct but it does seem to boot when I pass it 2 initrds.
I've implemented it by -initrd then all of the initrds after that as this was how I interpreted the qemu man page.
Fixes are in 37e75950ff.

poettering commented on src/vmspawn/vmspawn.c:
```diff
@@ -30,6 +32,7 @@
 #include "sd-event.h"
 #include "signal-util.h"
 #include "socket-util.h"
+#include "string-util-fundamental.h"
```
usually foo.h already imports foo-fundamental.h if such a pairing exists, and you sould only include foo.h hence.

or in other words, please change this to `#include "string-util.h"`, similar above

sam-leonard-ct replied:
Fixed in 37e75950ff.

poettering commented on src/vmspawn/vmspawn.c:
```diff
@@ -507,11 +521,63 @@ static int start_swtpm(sd_bus *bus, const char *scope, const char *swtpm, const
+         *   2. kernel stripped of suffix + .initrd
+         *   3. image + .initrd
+         */
+        initrd = strjoin(arg_qemu_kernel, ".initrd");
```
oom check

sam-leonard-ct replied:
Fixed in 37e75950ff.

poettering commented on src/vmspawn/vmspawn.c:
```diff
@@ -507,11 +521,63 @@ static int start_swtpm(sd_bus *bus, const char *scope, const char *swtpm, const
+        if (c)
+                *c = '\0';
+
+        initrd = strjoin(s, ".initrd");
```
oom check

sam-leonard-ct replied:
Fixed in 37e75950ff.

poettering commented on src/vmspawn/vmspawn.c:
```diff
@@ -507,11 +521,63 @@ static int start_swtpm(sd_bus *bus, const char *scope, const char *swtpm, const
+        }
+        if (errno != ENOENT)
+                return log_error_errno(errno, "Encountered error searching for initrd: %m");
+        initrd = mfree(initrd);
```
redundant line

sam-leonard-ct replied:
Fixed in 37e75950ff.

poettering commented on src/vmspawn/vmspawn.c:
```diff
@@ -507,11 +521,63 @@ static int start_swtpm(sd_bus *bus, const char *scope, const char *swtpm, const
+                return log_error_errno(errno, "Encountered error searching for initrd: %m");
+        initrd = mfree(initrd);
+
+        return -ESRCH;
```
should be ENOENT; no?

sam-leonard-ct replied:
Fixed in 37e75950ff.

poettering commented on src/vmspawn/vmspawn.c:
```diff
@@ -778,6 +846,22 @@ static int run_virtual_machine(void) {
+                initrd = strdup(arg_qemu_initrd);
+                if (!initrd)
+                        return log_oom();
+        } else if (arg_qemu_kernel && !endswith(arg_qemu_kernel, ".efi")) {
```
i think that endswith() check really should move into find_initrd()

sam-leonard-ct replied:
Fixed in 37e75950ff.

poettering commented on src/vmspawn/vmspawn.c:
```diff
@@ -1002,9 +1040,20 @@ static int run(int argc, char *argv[]) {
+
+        for (size_t i = 0; i < strv_length(names); i++) {
+                if (streq(names[i], "kvm"))
+                    kvm_device_fd = SD_LISTEN_FDS_START + i;
```
indentation

sam-leonard-ct replied:
Fixed in 56f995436f.

poettering commented on src/vmspawn/vmspawn.c:
```diff
@@ -1002,9 +1040,20 @@ static int run(int argc, char *argv[]) {
+        if (r < 0)
+                goto finish;
+
+        for (size_t i = 0; i < strv_length(names); i++) {
```
the return value of sd_listen_fds_with_names() already tells you the length of the array

sam-leonard-ct replied:
Fixed in 56f995436f.

poettering commented on src/vmspawn/vmspawn.c:
```diff
@@ -1002,9 +1040,20 @@ static int run(int argc, char *argv[]) {
+        for (size_t i = 0; i < strv_length(names); i++) {
+                if (streq(names[i], "kvm"))
+                    kvm_device_fd = SD_LISTEN_FDS_START + i;
+                if (streq(names[i], "vhost-vsock"))
```
maybe prefix with else

poettering commented on src/vmspawn/vmspawn.c:
```diff
@@ -1002,9 +1040,20 @@ static int run(int argc, char *argv[]) {
+                if (streq(names[i], "kvm"))
+                    kvm_device_fd = SD_LISTEN_FDS_START + i;
+                if (streq(names[i], "vhost-vsock"))
+                    vhost_device_fd = SD_LISTEN_FDS_START + i;
```
it would make sense to print a log message about fds that were passed in that we cannot make sense of, and close them (the log message should make it clear that they are closed now)

sam-leonard-ct replied:
Fixed in 56f995436f.

poettering commented on src/vmspawn/vmspawn.c:
```diff
@@ -585,15 +599,86 @@ static int find_initrd(char **ret_initrd) {
+        unit_name = strjoin(scope, "-virtiofsd", directory);
+        if (!unit_name)
+                return log_oom();
+
```
smae here, if there's only one parameter, maybe make this a regular templated services, rather than a transient one

sam-leonard-ct replied:
I'm not sure this is possible as it needs to know both the scope and the directory meaning there is more than one thing to be templated?

poettering replied:
can't the dir be derived from %i?

sam-leonard-ct replied:
I don't see how.
because of `--bind` we need multiple virtiofsd instances in the same scope.
@richardmaw-codethink has also pointed out that this limits you to only mounting a directory once with virtiofsd, preventing multiple VMs from sharing a read only directory image.

sam-leonard-ct replied:
Another thing to note is that this limits the lengths of paths which can be used, `UNIT_NAME_MAX` (255) is considerably less than `PATH_MAX` (4096).

poettering commented on src/vmspawn/vmspawn.c:
```diff
@@ -105,6 +106,10 @@ static int help(void) {
+               "%3$sUser Namespacing:%4$s\n"
+               "     --private-users=UIDBASE[:NUIDS]\n"
+               "                            Configure the UID/GID range to map into the\n"
+               "                            virtiofsd namespace\n\n"
```
i am not sure i grok this. what dos this do?

sam-leonard-ct replied:
Currently mkosi builds images inside of a user namespace, which means the resultant directory image is owned by PID 0 in the user's subuid namespace.
In order to boot an image like this I need to tell virtiofsd to map the external user namespace UIDs into the correct UIDs for the VM.
While discussing with Daan he asked that I call this --private-users. I agree that the help message isn't very helpful though.
The man page has a much more thorough explanation and gives a hint on when to use it under the part for `-D`

YHNdnzj commented on src/vmspawn/vmspawn-mount.c:
```diff
@@ -0,0 +1,63 @@
+#include "path-util.h"
+#include "string-util.h"
+
+void runtime_mount_free_all(RuntimeMount *mounts, size_t n) {
+        assert(mounts || n == 0);
+
+        FOREACH_ARRAY(m, mounts, n) {
+                free(m->source);
+                free(m->target);
+        }
+
+        free(mounts);
+}
```
You can use my rework on `MachineCredential` as an example: bd546b9b48e575dbea37ac9c7cf5e79985712e54

sam-leonard-ct replied:
Ooh thanks thats definitely nicer to work with.
Also thanks for commenting this I did a rebase and got very lost as to where arg_n_credentials had gone 😂.

sam-leonard-ct replied:
Fixed in 56f995436f.

sam-leonard-ct commented on src/vmspawn/vmspawn-scope.c:
```diff
@@ -0,0 +1,383 @@
+
+/* NOTE: this function should really take cmdline as `const char * const *` but cannot until sd_bus_message_append_str does as well */
+int run_command_bound_to_scope(sd_bus *bus, const char *scope, const char *service_name, char **cmdline,
+                               char **cleanup_cmdline, char **extra_properties) {
```
Fixed in 56f995436f.

YHNdnzj commented on src/vmspawn/vmspawn-util.c:
```diff
@@ -33,6 +33,22 @@ OvmfConfig* ovmf_config_free(OvmfConfig *config) {
         return mfree(config);
 }
 
+int parse_qemu_network_stack(const char *s, QemuNetworkStack *net_stack) {
```
Just make this return `QemuNetworkStack`? `_QEMU_NET_INVALID` can be returned if no match is found

sam-leonard-ct replied:
Yeah this is a lot nicer, just didn't occur to me at the time :)

YHNdnzj commented on src/vmspawn/vmspawn-util.c:
```diff
@@ -33,6 +33,22 @@ OvmfConfig* ovmf_config_free(OvmfConfig *config) {
+
+        if (streq(optarg, "user"))
+                *net_stack = QEMU_NET_USER;
+        else if (streq(optarg, "tap"))
```
Then, you can use early return rather than `else if`s

sam-leonard-ct commented:
Added some code to read the image's partition table and synthesise a root= kernel argument.
As requested by Daan here: https://github.com/systemd/mkosi/pull/2037#discussion_r1431319083

sam-leonard-ct commented:
Pushed a minor whitespace change to the help message.

poettering commented on src/vmspawn/vmspawn-scope.c:
```diff
@@ -0,0 +1,426 @@
+        assert(scope);
+        assert(cmdline);
+
+        r = unit_name_mangle_with_suffix(service_name, "as service", 0, ".service", &unit_name);
```
the "mangle" api is primarily there to turn slightly icky user input into valid unit names. but when we create unit names internally from our own components, then these should be clean, and not require mangling. hence, please use strjoin() and friends.

poettering commented on src/vmspawn/vmspawn-scope.c:
```diff
@@ -0,0 +1,426 @@
+                                  "Description", "s", description,
+                                  "AddRef", "b", 1,
+                                  "BindsTo", "as", 1, scope,
+                                  "CollectMode", "s",     "inactive-or-failed");
```
whitespace looks a bit unexpected here

poettering commented on src/vmspawn/vmspawn-scope.c:
```diff
@@ -0,0 +1,426 @@
+        if (r < 0)
+                return log_error_errno(r, "Could not watch job: %m");
+
+        r = unit_name_mangle_with_suffix(unit_name, "as socket", 0, ".socket", &socket_name);
```
as above

sam-leonard-ct replied:
Fixed in eae47ee9d9.

poettering commented on src/vmspawn/vmspawn-scope.c:
```diff
@@ -0,0 +1,426 @@
+                break;
+        default:
+                return log_error_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "Invalid socket type: %d", socket_type);
+        }
```
can you move the translation of the socket type to some helper please? i.e. one of those table-based xyz_to_string( )helpers

sam-leonard-ct replied:
I found a lookup table for SOCK_* entries in socket-util.c so I'm going to move to using that as I think there is no need for a function once its a table lookup.

sam-leonard-ct replied:
Fixed in eae47ee9d9.

poettering commented on src/vmspawn/vmspawn.c:
```diff
@@ -127,6 +131,7 @@ static int parse_argv(int argc, char *argv[]) {
                 { "qemu-kvm",        required_argument, NULL, ARG_QEMU_KVM        },
                 { "qemu-vsock",      required_argument, NULL, ARG_QEMU_VSOCK      },
                 { "vsock-cid",       required_argument, NULL, ARG_VSOCK_CID       },
+                { "vtpm",            required_argument, NULL, ARG_VTPM            },
```
i'd just call this "tpm" btw. the fact that it is virtual is kinda implied i'd say

sam-leonard-ct replied:
Fixed in eae47ee9d9.

poettering commented on src/vmspawn/vmspawn.c:
```diff
@@ -540,6 +551,10 @@ static int run_virtual_machine(void) {
         if (!machine)
                 return log_oom();
 
+
```
spurious newline

sam-leonard-ct replied:
Fixed in eae47ee9d9.

poettering commented on src/vmspawn/vmspawn.c:
```diff
@@ -540,6 +551,10 @@ static int run_virtual_machine(void) {
 
+
+        if (arg_kernel && access(arg_kernel, F_OK) < 0)
+                return log_error_errno(errno, "Kernel not found at %s: %m", arg_kernel);
```
i don't think we need to do this check here. we'll notice soon enough, no? qemu will fail if it cannot find the path

sam-leonard-ct replied:
Fixed in eae47ee9d9.

poettering commented on src/vmspawn/vmspawn.c:
```diff
@@ -551,10 +633,13 @@ static int run_virtual_machine(void) {
 
+        STRV_FOREACH(i, arg_initrds)
+                if (access(*i, F_OK) < 0)
+                        return log_error_errno(errno, "Initrd not found at %s: %m", *i);
```
same here

sam-leonard-ct replied:
Fixed in eae47ee9d9.

poettering commented on man/systemd-vmspawn.xml:
```diff
@@ -165,6 +165,18 @@
         </varlistentry>
 
+        <varlistentry>
+          <term><option>--qemu-net=</option><replaceable>user|tap|none</replaceable></term>
```
i wonder if this wouldn't be better to default to off, i.e. `none`

sam-leonard-ct replied:
Now defaults to none in eae47ee9d9.

poettering commented on src/vmspawn/vmspawn.c:
```diff
@@ -259,6 +264,12 @@ static int parse_argv(int argc, char *argv[]) {
 
+                case ARG_QEMU_NET:
+                        arg_qemu_net = parse_qemu_network_stack(optarg);
+                        if (arg_qemu_net == _QEMU_NET_INVALID)
```
usually we just do a `arg_qemu_net < 0` check here, to catch all kinds of errors

sam-leonard-ct replied:
Fixed in eae47ee9d9.

poettering commented on src/vmspawn/vmspawn.c:
```diff
@@ -687,6 +688,30 @@ static int run_virtual_machine(void) {
                 return log_oom();
 
+        if (use_kvm && kvm_device_fd > 0) {
+                accel = "kvm,device=/dev/fdset/1";
```
hmm what's /dev/fdset/? That doesn't exist here. Is that a magic string for qemu?

sam-leonard-ct replied:
This tells qemu to receive the KVM device from the first fd set, we provide a fd to `/dev/kvm` in set one if we are provided one from `sd_listen_fds`, so this argument is required so that qemu uses it instead of trying to open `/dev/kvm` itself.

poettering replied:
can you add a comment that /dev/fdset/ is a magic string for qemu and not an actual fs path then

sam-leonard-ct replied:
Fixed in eae47ee9d9.

poettering commented on src/vmspawn/vmspawn.c:
```diff
@@ -1025,9 +1063,25 @@ static int run(int argc, char *argv[]) {
+                        vhost_device_fd = fd;
+                else {
+                        log_notice("Couldn't recognise passed fd %d (%s), closing fd and ignoring...", fd, names[i]);
+                        close(fd);
```
safe_close()

sam-leonard-ct replied:
Fixed in eae47ee9d9.

poettering commented on src/vmspawn/vmspawn.c:
```diff
@@ -601,15 +615,86 @@ static int find_initrd(char **ret_initrd) {
+                return log_error_errno(r, "Error while searching for virtiofsd: %m");
+
+        if (!virtiofsd) {
+                FOREACH_STRING(dir, "/usr/libexec/virtiofsd", "/usr/libexec/virtiofsd") {
```
hmm, which distro puts the binary there?

poettering replied:
the two paths are exacty the same?

sam-leonard-ct replied:
I'm not sure, this was behaviour inherited from mkosi: https://github.com/systemd/mkosi/blob/a4444f01d0e29e380c576cf218f3ef3dc66b8409/mkosi/qemu.py#L302-L312

sam-leonard-ct replied:
And on the second point this is a typo and the second one should be `/usr/lib/virtiofsd`

DaanDeMeyer replied:
@poettering Fedora puts it there

sam-leonard-ct replied:
@DaanDeMeyer is the version fedora puts there the Rust rewrite?

On my Ubuntu install it is still on the pre rust-rewrite version and its located at `/usr/lib/qemu/virtiofsd`.
As in should I include `/usr/lib/qemu/virtiofsd` as well as a fallback path.

The issue is that the arguments aren't compatible and it requires root so finding the old version wouldn't be desirable unless there is potential for the rust-rewrite version to appear there in future.

DaanDeMeyer replied:
@sam-leonard-ct We should only support the rust rewrite

sam-leonard-ct replied:
Fixed in eae47ee9d9.

poettering commented on src/vmspawn/vmspawn.c:
```diff
@@ -601,15 +615,86 @@ static int find_initrd(char **ret_initrd) {
+
+        if (!virtiofsd) {
+                FOREACH_STRING(dir, "/usr/libexec/virtiofsd", "/usr/libexec/virtiofsd") {
+                        if (access(dir, X_OK) >= 0) {
```
is this a file or a dir? you named the variable "dir", but afaics this actually the binary file?

sam-leonard-ct replied:
Yep these are files, I will change the name.

sam-leonard-ct replied:
Fixed in eae47ee9d9.

poettering commented on src/vmspawn/vmspawn-mount.c:
```diff
@@ -0,0 +1,64 @@
+        if (r < 0)
+                return r;
+
+        /* virtiofsd only supports directories */
```
why check this here though? i mean, it will fail on its own in this case, no? no need to check this ahead of time?

sam-leonard-ct replied:
Without this check you get a very unintuitive and confusing error:
![image](https://github.com/systemd/systemd/assets/145554582/a45e2668-2bf4-4ec1-a4d4-9f6d85a8b6f1)

with it we fail with a clear error message explaining what went wrong:
![image](https://github.com/systemd/systemd/assets/145554582/5d362790-2f86-47e4-9e2b-5c6b578f7c1c)

In my opinion the failure on qemu's side is confusing enough to warrant us checking.

poettering commented on src/vmspawn/vmspawn.c:
```diff
@@ -760,6 +763,117 @@ static int start_virtiofsd(sd_bus *bus, const char *scope, const char *directory
+        }
+
+        return -ENOENT;
+}
```
please use dissect_image_file_and_warn() for this, and then use the data it returns. we want only a single dissector for this.

i.e. it will return a DissectedImage object to you, and you can get the part uuid from it via `.partitions[PARTITION_ROOT].uuid`.

sam-leonard-ct replied:
Fixed in eae47ee9d9.

poettering commented:
looks good, some points, see above

sam-leonard-ct commented on src/vmspawn/vmspawn.c:
```diff
@@ -601,15 +615,86 @@ static int find_initrd(char **ret_initrd) {
+        if (r < 0)
+                return log_error_errno(r, "Failed to find 'rm' binary: %m");
+
+        r = mkdtemp_malloc("/tmp/vmspawn-virtiofsd-XXXXXX", &state);
```
this is nice, I'm going to try to respect XDG_RUNTIME_DIR for this as these should be user units.

poettering replied:
use $RUNTIME_DIRECTORY which is what RuntimeDirectory= initializes and works the same for per-user and system services

sam-leonard-ct replied:
This change is now in eae47ee9d9.

poettering commented on src/vmspawn/vmspawn-scope.c:
```diff
@@ -0,0 +1,426 @@
+        if (r < 0)
+                return log_error_errno(r, "Could not watch job: %m");
+
+        if (asprintf(&unmangled, "sd-vmspawn-%"PRIu32"-%s", random_u32(), machine_name) < 0)
```
containers are simply prefixed with `machine-` btw. i'd stick to the same naming scheme here

sam-leonard-ct replied:
Fixed in eae47ee9d9.

poettering commented on src/vmspawn/vmspawn-scope.c:
```diff
@@ -0,0 +1,426 @@
+        if (asprintf(&unmangled, "sd-vmspawn-%"PRIu32"-%s", random_u32(), machine_name) < 0)
+                return log_oom();
+
+        r = unit_name_mangle_with_suffix(unmangled, "as scope name", 0, ".scope", &scope);
```
as mentioned elsewhere, please don't use the "mangle" apis except when processing direct user input

sam-leonard-ct replied:
Fixed in eae47ee9d9.

poettering commented on src/vmspawn/vmspawn-util.c:
```diff
@@ -33,6 +33,19 @@ OvmfConfig* ovmf_config_free(OvmfConfig *config) {
+        if (streq(s, "none"))
+                return QEMU_NET_NONE;
+
+        return _QEMU_NET_INVALID;
```
just use the usual xyz_from_string() logic we have for this, i.e. string-table.h

sam-leonard-ct replied:
Fixed in eae47ee9d9.

DaanDeMeyer commented on src/vmspawn/vmspawn.c:
```diff
@@ -668,6 +702,24 @@ static int start_virtiofsd(sd_bus *bus, const char *scope, const char *directory
                 return log_oom();
 
+        if (arg_uid_shift != UID_INVALID) {
+                r = strv_extend(&cmdline, "--uid-map");
```
So since these options will make virtiofsd shell out to `newuidmap/newgidmap` I'm wondering if we should instead allow passing a userns fd to virtiofsd via `sd_listen()`. That way mkosi can take care of passing in a user namespace and vmspawn would run virtiofsd in that user namespace.

It'll also be required if we want to implement `--ephemeral` since we'll need to join the userns to be able to copy the files from the root directory.

sam-leonard-ct replied:
can user namespaces be passed to units?

sam-leonard-ct commented on src/vmspawn/vmspawn.c:
```diff
@@ -427,20 +557,302 @@
+        assert(our_runtime_dir);
+        assert(ret_state_tempdir);
+
+        r = create_runtime_tempdir(our_runtime_dir, "tpm-XXXXXX", &state_dir, &state_runtime_dir);
```
oops

sam-leonard-ct replied:
Should be fixed now in b2efd0d301.

DaanDeMeyer commented on src/vmspawn/vmspawn-scope.c:
```diff
@@ -0,0 +1,413 @@
+        /*
+         * Starts a socket/service unit pair bound to the given scope.
+         *
+         * The socket unit listens on `socket_path` with the for connections of type `socket_type`.
```
Typo

DaanDeMeyer commented on src/vmspawn/vmspawn-scope.c:
```diff
@@ -0,0 +1,413 @@
+         * Starts a socket/service unit pair bound to the given scope.
+         *
+         * The socket unit listens on `socket_path` with the for connections of type `socket_type`.
+         * When a connection to the socket is initiated the service is started to and passed the socket's file descriptor as fd 3.
```
to => as well

DaanDeMeyer commented on src/vmspawn/vmspawn.c:
```diff
@@ -474,20 +488,124 @@ static int cmdline_add_vsock(char ***cmdline, int vsock_fd) {
+        /* Creates a new temporary directory under using the template
+         *
+         * ret_tempdir is the full path to the tempdir
+         * ret_runtime_directory_property is the tempdir formatted as
```
Sentence cut short

DaanDeMeyer commented on src/vmspawn/vmspawn.c:
```diff
@@ -474,20 +488,124 @@ static int cmdline_add_vsock(char ***cmdline, int vsock_fd) {
+        assert(our_runtime_dir);
+        assert(template);
+
+        /* Creates a new temporary directory under using the template
```
using the given template

DaanDeMeyer commented on src/vmspawn/vmspawn.c:
```diff
@@ -474,20 +488,124 @@ static int cmdline_add_vsock(char ***cmdline, int vsock_fd) {
         int r;
-        _cleanup_close_ int vsock_fd = -EBADF;
+
+        if (getuid() == 0)
+                r = sd_bus_open_system(&bus);
+        else
+                r = sd_bus_open_user(&bus);
+        if (r < 0)
+                return log_error_errno(r, "Failed to connect to systemd bus: %m");
```
sd_bus_default()?

sam-leonard-ct replied:
I believe that this is actually the correct way of doing this, without it vmspawn cannot open a bus connection when invoked as root.

DaanDeMeyer commented on src/vmspawn/vmspawn.c:
```diff
@@ -538,6 +656,32 @@ static int run_virtual_machine(void) {
 
+        /* if we are going to be starting any units with state then create our runtime dir */
+        if (arg_tpm != 0) {
+                _cleanup_free_ char *runtime_directory = NULL;
+                const char *e = getenv("RUNTIME_DIRECTORY");
+                if (e)
+                        runtime_directory = strdup(e);
+                else if (getuid() == 0)
+                        runtime_directory = strdup("/run");
+                else {
+                        r = xdg_user_runtime_dir(&runtime_directory, "");
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to find user's runtime directory: %m");
+                }
+                if (!runtime_directory)
+                        return log_oom();
+
+                /* ensure runtime_dir/systemd/vmspawn exists */
+                our_runtime_dir = path_join(runtime_directory, "systemd/vmspawn");
+                if (!our_runtime_dir)
+                        return log_oom();
+
+                r = mkdir_p(our_runtime_dir, 0755);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to create runtime directory: %m");
```
This should be a helper function (maybe in daemon-util.h?) There's nothing vmspawn specific about this.

sam-leonard-ct replied:
I think this might be more appropriate in path-lookup.[ch]?

sam-leonard-ct replied:
daemon-util seems to be more focussed on notify sd-notify things.

DaanDeMeyer replied:
Yeah maybe that's better

DaanDeMeyer commented on src/vmspawn/vmspawn.c:
```diff
@@ -685,6 +829,50 @@ static int run_virtual_machine(void) {
+                        return log_oom();
+
+                if (native_architecture() == ARCHITECTURE_X86_64)
+                        r = strv_extend_strv(&cmdline, STRV_MAKE("-device", "tpm-tis,tpmdev=tpm0"), /* filter_duplicates= */ false);
```
strv_extend_many()

DaanDeMeyer commented on src/vmspawn/vmspawn.c:
```diff
@@ -685,6 +829,50 @@ static int run_virtual_machine(void) {
+                if (native_architecture() == ARCHITECTURE_X86_64)
+                        r = strv_extend_strv(&cmdline, STRV_MAKE("-device", "tpm-tis,tpmdev=tpm0"), /* filter_duplicates= */ false);
+                else if (IN_SET(native_architecture(), ARCHITECTURE_ARM64, ARCHITECTURE_ARM64_BE))
+                        r = strv_extend_strv(&cmdline, STRV_MAKE("-device", "tpm-tis-device,tpmdev=tpm0"), /* filter_duplicates= */ false);
```
Ditto

DaanDeMeyer commented on src/vmspawn/vmspawn.c:
```diff
@@ -685,6 +829,50 @@ static int run_virtual_machine(void) {
+                if (r < 0)
+                        return log_oom();
+
+                r = strv_extend_strv(&cmdline, STRV_MAKE("-tpmdev", "emulator,id=tpm0,chardev=chrtpm"),
```
strv_extend_many()

DaanDeMeyer commented on src/vmspawn/vmspawn.c:
```diff
@@ -474,20 +488,124 @@ static int cmdline_add_vsock(char ***cmdline, int vsock_fd) {
+        if (r < 0)
+                return log_oom();
+
+        r = strv_extend_strv(&cmdline, STRV_MAKE("--ctrl", "type=unixio,fd=3"), /* filter_duplicates= */ false);
```
strv_extend_many()

DaanDeMeyer commented on src/vmspawn/vmspawn-scope.h:
```diff
@@ -0,0 +1,12 @@
+#include "sd-bus.h"
+
+int start_transient_scope(sd_bus *bus, const char *machine_name, bool allow_pidfd, char **ret_scope);
+int run_command_bound_to_scope(sd_bus *bus, const char *scope, const char *service_name, char **cmdline,
+                               char **cleanup_cmdline, char **extra_properties);
```
This function should be introduced in either the commit that makes use of it or in a separate commit.

sam-leonard-ct replied:
currently it isn't used should I split out or just remove it, it was what I had originally written but it seems like socket/service units are just more useful for vmspawn.

DaanDeMeyer replied:
@sam-leonard-ct Please remove it if not used. You can keep it around in a private branch if you think you might need it in the future

DaanDeMeyer commented on src/vmspawn/vmspawn-scope.c:
```diff
@@ -0,0 +1,413 @@
+        if (r < 0)
+                return bus_log_parse_error(r);
+
+        r = bus_wait_for_jobs_one(w, object, /* quiet */ false, NULL);
+        if (r < 0)
+                return r;
+
+        return 0;
```
Can be simplified to `return bus_wait_for_jobs_one()`

DaanDeMeyer commented on src/vmspawn/vmspawn.c:
```diff
@@ -90,6 +92,7 @@ static int help(void) {
                "     --qemu-vsock=BOOL      Configure whether to use qemu with a vsock or not\n"
                "     --vsock-cid=           Specify the CID to use for the qemu guest's vsock\n"
                "     --tpm=BOOL             Configure whether to use a virtual TPM or not\n"
+               "     --kernel=PATH          Specify the kernel for direct kernel boot\n"
```
Maybe we should name this --linux= instead of --kernel to more clearly indicate we're talking about a linux kernel image here.

DaanDeMeyer commented on src/vmspawn/vmspawn.c:
```diff
@@ -301,9 +312,9 @@ static int parse_argv(int argc, char *argv[]) {
-                strv_free(arg_parameters);
-                arg_parameters = strv_copy(argv + optind);
-                if (!arg_parameters)
+                strv_free(arg_kernel_cmdline_extra);
```
Why do we have to free the strv here? It should be `NULL` at this point no?

sam-leonard-ct replied:
I have wondered this too, nspawn did it so I thought it safer to include as well.

DaanDeMeyer commented on man/systemd-vmspawn.xml:
```diff
@@ -85,52 +85,52 @@
 
-      <varlistentry>
-        <term><option>--qemu-kvm=</option><replaceable>BOOL</replaceable></term>
+        <varlistentry>
```
The indentation changes here should be a separate commit

DaanDeMeyer commented on man/systemd-vmspawn.xml:
```diff
@@ -142,38 +142,54 @@
 
-        <listitem><para>Start QEMU in graphical mode.</para>
+          <listitem>
+            <para>Set the kernel initrd to use for direct kernel boot.</para>
```
kernel initrd => initrd

DaanDeMeyer commented on man/systemd-vmspawn.xml:
```diff
@@ -142,38 +142,54 @@
-      <varlistentry>
-        <term><option>--qemu-gui</option></term>
+        <varlistentry>
+          <term><option>--initrd=</option><replaceable>PATH</replaceable></term>
```
Are we sure that qemu merges multiple initrds if more than one is specified?

sam-leonard-ct replied:
I'm not super sure, it doesn't seem to break.
It was proposed here: https://github.com/systemd/systemd/pull/29960#discussion_r1425688080

sam-leonard-ct replied:
To reiterate here, it does not merge initrds, we must do this ourself if we want this feature. (I think this should be considered in the scope of whatever is building the image not necessarily in vmspawn's scope.)

DaanDeMeyer commented on src/vmspawn/vmspawn.c:
```diff
@@ -596,6 +616,68 @@ static int start_tpm(sd_bus *bus, const char *scope, const char *our_runtime_dir
+        if (endswith(arg_kernel, ".efi")) {
+                ret_initrd = NULL;
+                return 0;
+        }
```
This should use `inspect_kernel()` to figure out the type instead of checking the file extension

DaanDeMeyer replied:
To be fair I would drop this automatic initrd finding logic completely for now and just require the initrd to be specified explicitly.

sam-leonard-ct replied:
This certainly makes it a lot simpler 😂

DaanDeMeyer commented on src/vmspawn/vmspawn.c:
```diff
@@ -775,6 +785,22 @@ static int run_virtual_machine(void) {
+        case QEMU_NET_USER:
+                r = strv_extend_strv(&cmdline, STRV_MAKE("-nic", "user,model=virtio-net-pci"), /* filter_duplicates= */ false);
+                break;
+        case QEMU_NET_TAP:
```
We should bail if we're not running as root and this is selected as qemu will just fail with a permission error

DaanDeMeyer replied:
We should also mention in the man page that this requires a correctly configured systemd-networkd to be running on the host so the host interface gets configured correctly.

poettering replied:
strv_extend_many()

DaanDeMeyer commented on src/vmspawn/vmspawn.c:
```diff
@@ -1165,9 +1203,25 @@ static int run(int argc, char *argv[]) {
+                else if (streq(names[i], "vhost-vsock"))
+                        vhost_device_fd = fd;
+                else {
+                        log_notice("Couldn't recognise passed fd %d (%s), closing fd and ignoring...", fd, names[i]);
```
"recognise" => "recognize"

DaanDeMeyer commented on src/vmspawn/vmspawn.c:
```diff
@@ -801,6 +799,32 @@ static int run_virtual_machine(void) {
+        else
+                accel = "tcg";
+
+        r = strv_extend_strv(&cmdline, STRV_MAKE("-accel", accel), /* filter_duplicates= */ false);
```
strv_extend_many()

DaanDeMeyer commented on src/vmspawn/vmspawn.c:
```diff
@@ -801,6 +799,32 @@ static int run_virtual_machine(void) {
                 return log_oom();
 
+        if (use_kvm && kvm_device_fd > 0) {
+                /* /dev/fdset/1 is magic string to tell qemu where to find the fd for /dev/kvm
```
"is a magic string"

DaanDeMeyer commented on src/vmspawn/vmspawn.c:
```diff
@@ -731,20 +732,17 @@ static int run_virtual_machine(void) {
 
+        if (IN_SET(native_architecture(), ARCHITECTURE_ARM64, ARCHITECTURE_ARM64_BE))
+                machine = "type=virt";
+        else
```
We should only apply this on x86 and fail for other architectures as we don't know the machine to use in that case

sam-leonard-ct replied:
https://github.com/systemd/mkosi/blob/a503c80dca9c7f569d9333f89f85d3f5ada60859/mkosi/config.py#L368-L384
Would it not be more correct to implement something like this from mkosi?

DaanDeMeyer commented on src/vmspawn/vmspawn.c:
```diff
@@ -689,15 +690,15 @@ static int find_initrd(char **ret_initrd) {
+        _cleanup_free_ char *qemu_binary = NULL, *mem = NULL, *trans_scope = NULL, *our_runtime_dir = NULL;
+        _cleanup_close_ int notify_sock_fd = -EBADF;
         _cleanup_strv_free_ char **cmdline = NULL;
+        _cleanup_free_ int *pass_fds = NULL;
```
This variable doesn't seem to be actually passed to the qemu child process anywhere?

sam-leonard-ct replied:
The variable itself isn't but the FDs it stores are.
I use this variable to track which fds shouldn't be closed by safe_fork essentially.

poettering commented:
btw, my recommendation would be to split stuff out some of these commits, if they are easy to split out, and we can merge them quicker. I do that all the time with my PRs. If things make sense on their own it's way easier to review+merge small PRs than those with 10+. And if it's in it's in. Then rebase immediately and reiterate.

doing this is not a requirement, but it will grease the wheels to get things merged more quickly, because experience tells me large PRs scare reviewers ;-)

sam-leonard-ct commented:
Fixed all of the review comments, as mentioned I'm going to split out the virtiofsd/directory stuff from this PR into its own one to make this one easier to merge.

sam-leonard-ct commented:
to faciliate this I'm going to move "synthesise root= ..." to before the virtiofsd commits.

DaanDeMeyer commented on src/vmspawn/vmspawn-scope.h:
```diff
@@ -0,0 +1,12 @@
+#include "sd-bus.h"
+
+int start_transient_scope(sd_bus *bus, const char *machine_name, bool allow_pidfd, char **ret_scope);
+int run_command_bound_to_scope(sd_bus *bus, const char *scope, const char *service_name, char **cmdline,
```
This still needs to be removed

DaanDeMeyer commented on src/vmspawn/vmspawn-scope.c:
```diff
@@ -0,0 +1,306 @@
+        assert(unit_name_prefix);
+        assert(socket_path);
+        assert(cmdline);
+        assert_return(IN_SET(socket_type, SOCK_STREAM, SOCK_SEQPACKET, SOCK_DGRAM), -EOPNOTSUPP);
```
We don't really use `assert_return()` in internal functions. Please make this a regular if and log an error if it fails.

sam-leonard-ct replied:
I'm not sure we actually need the assertion at all, theres nothing wrong with the other socket types and I think this was an artifact of how I was turning it into a string. As I now use a table for this it I think we can drop the assert?

DaanDeMeyer commented on src/vmspawn/vmspawn.c:
```diff
@@ -84,12 +89,13 @@ static int help(void) {
                "     --firmware=PATH|list   Select firmware definition file (or list available)\n"
                "\n%3$sSystem Identity:%4$s\n"
-               "  -M --machine=NAME         Set the machine name for the container\n"
+               "  -M --machine=NAME         Set the machine name for the virtual machine\n"
```
Should be a separate commit

DaanDeMeyer commented on src/vmspawn/vmspawn.c:
```diff
@@ -685,6 +811,49 @@ static int run_virtual_machine(void) {
+                if (r < 0)
+                        return log_oom();
+
+                if (native_architecture() == ARCHITECTURE_X86_64)
+                        r = strv_extend_many(&cmdline, "-device", "tpm-tis,tpmdev=tpm0");
+                else if (IN_SET(native_architecture(), ARCHITECTURE_ARM64, ARCHITECTURE_ARM64_BE))
+                        r = strv_extend_many(&cmdline, "-device", "tpm-tis-device,tpmdev=tpm0");
```
We should fail if user's try to do tpm with another architecture that isn't one of these

DaanDeMeyer commented on src/vmspawn/vmspawn.c:
```diff
@@ -595,6 +598,32 @@ static int start_tpm(sd_bus *bus, const char *scope, const char *our_runtime_dir
+        else if (image->partitions[PARTITION_USR].found)
+                root = strjoin("mount.usr=PARTUUID=", SD_ID128_TO_UUID_STRING(image->partitions[PARTITION_USR].uuid));
+        else
+                return -ENOENT;
```
Log an error message here

DaanDeMeyer commented on src/vmspawn/vmspawn.c:
```diff
@@ -790,6 +819,19 @@ static int run_virtual_machine(void) {
+                        _cleanup_free_ char *root = NULL;
+                        r = finalize_root(&root);
+                        if (r < 0)
+                                return log_error_errno(r, "Cannot perform a direct kernel boot without a root or usr partition: %m");
```
Move this error message into `finalize_root()`

DaanDeMeyer commented on src/vmspawn/vmspawn.c:
```diff
@@ -790,6 +819,19 @@ static int run_virtual_machine(void) {
+                /* We can't rely on gpt-auto-generator when direct kernel booting so synthesize a root=
+                 * kernel argument instead. */
+                if (arg_image && !strv_find_startswith(arg_kernel_cmdline_extra, "root=")) {
+                        _cleanup_free_ char *root = NULL;
```
Missing newline after declarations

sam-leonard-ct commented:
Fixed all of Daan's comments and I'm going to go over all of the ones from this PR (https://github.com/systemd/systemd/pull/31076) that are about content from this PR now.

sam-leonard-ct commented:
I have now completed all of the review comments from https://github.com/systemd/systemd/pull/31076 which pertain to content from this PR and am going to start working on the comments only relevant to that PR now.

sam-leonard-ct commented:
Updated start_tpm to use path_join instead of strjoin.

sam-leonard-ct commented:
Fixed more review comments from here: https://github.com/systemd/systemd/pull/31076

sam-leonard-ct commented:
Removed `socket_desc` and changed the description to just `p->listen_address`, this is because the logs ended up reading Listening on Listening on <address>. This changes means it just reads Listening on <address> which is what I would expect to read.

sam-leonard-ct commented:
Replaced a strv_extend with a strv_consume.

sam-leonard-ct commented:
Fixed more review comments from here: https://github.com/systemd/systemd/pull/31076
